src/main/java/com/epam/aidial/core/util/ProxyUtil.java

```diff
@@ -161,50 +161,80 @@ private static void collectAttachedFilesCustomInput(JsonNode input, Consumer<Str
+            JsonNode urlNode = imageNode.get("url");
+            collectAttachedFilesFromUrl(urlNode, null, consumer);
+        }
+    }
+
+    private static void collectAttachedFilesFromCustomContent(JsonNode customContent, Consumer<String> consumer) {
+        if (customContent == null) {
+            return;
+        }
+        ArrayNode attachments = (ArrayNode) customContent.get("attachments");
+        if (attachments != null) {
+            for (int i = 0; i < attachments.size(); i++) {
+                JsonNode attachment = attachments.get(i);
+                collectAttachedFile(attachment, consumer);
+            }
+        }
+        ArrayNode stages = (ArrayNode) customContent.get("stages");
+        if (stages != null) {
+            for (int i = 0; i < stages.size(); i++) {
+                JsonNode stage = stages.get(i);
+                attachments = (ArrayNode) stage.get("attachments");
+                if (attachments == null) {
+                    continue;
+                }
                 for (int j = 0; j < attachments.size(); j++) {
                     JsonNode attachment = attachments.get(j);
                     collectAttachedFile(attachment, consumer);
                 }
             }
-            ArrayNode stages = (ArrayNode) customContent.get("stages");
-            if (stages != null) {
-                for (int j = 0; j < stages.size(); j++) {
-                    JsonNode stage = stages.get(j);
-                    attachments = (ArrayNode) stage.get("attachments");
-                    if (attachments == null) {
-                        continue;
-                    }
-                    for (int k = 0; k < attachments.size(); k++) {
-                        JsonNode attachment = attachments.get(k);
-                        collectAttachedFile(attachment, consumer);
-                    }
-                }
-            }
         }
     }
 
-    public static void collectAttachedFile(JsonNode attachment, Consumer<String> consumer) {
-        JsonNode urlNode = attachment.get("url");
+    private static void collectAttachedFilesChatCompletion(ObjectNode tree, Consumer<String> consumer) {
+        ArrayNode messages = (ArrayNode) tree.get("messages");
+        if (messages == null) {
+            return;
+        }
+        for (int i = 0; i < messages.size(); i++) {
+            JsonNode message = messages.get(i);
+            JsonNode content = message.get("content");
+            collectAttachedFilesFromContent(content, consumer);
+
+            JsonNode customContent = message.get("custom_content");
+            collectAttachedFilesFromCustomContent(customContent, consumer);
+        }
+    }
+
+    private static void collectAttachedFilesFromUrl(JsonNode urlNode, JsonNode typeNode, Consumer<String> consumer) {
         if (urlNode == null) {
             return;
         }
 
         String url = urlNode.textValue();
 
-        JsonNode typeNode = attachment.get("type");
+        if (url == null) {
```

Comment:
the check could be unnecessary.

Reply:
JsonNode is null means that the field is missing from the dictionary.
But it may be not null if the field is present but equals to JSON `null`.

So it's required for this case:

```json
{"type": "image_url", "image_url": {"url": null}},
```

If the check is missing, this will be called: `consumer.accept(null);`.